src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php

```diff
@@ -300,6 +289,444 @@ public static function analyze(
+                                    $valid_ints[] = new TLiteralInt((int) $sub_atomic_type->value);
+                                } elseif ($sub_atomic_type instanceof TNumericString) {
+                                    $castable_types[] = new TInt();
+                                } else {
+                                    // see above, why this is less strict
+                                    $castable_types[] = new TInt();
+                                }
+                            }
+
+                            continue 2;
+                        }
+                    }
+
+                    if ($intersection_type instanceof TObjectWithProperties
+                        && isset($intersection_type->methods['__toString'])
+                    ) {
+                        $castable_types[] = new TInt();
+
+                        continue 2;
+                    }
+                }
+            }
+
+            if ($atomic_type instanceof TNonEmptyArray
+                || $atomic_type instanceof TNonEmptyList
+            ) {
+                $risky_cast[] = $atomic_type->getId();
+
+                $valid_ints[] = new TLiteralInt(1);
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TArray
+                || $atomic_type instanceof TList
+                || $atomic_type instanceof TKeyedArray
+            ) {
+                // if type is not specific, it can be both 0 or 1, depending on whether the array has data or not
+                // welcome to off-by-one hell if that happens :-)
+                $risky_cast[] = $atomic_type->getId();
+
+                $valid_ints[] = new TLiteralInt(0);
+                $valid_ints[] = new TLiteralInt(1);
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TTemplateParam) {
+                $atomic_types = array_merge($atomic_types, $atomic_type->as->getAtomicTypes());
+
+                continue;
+            }
+
+            // always 1 for "error" cases
+            $valid_ints[] = new TLiteralInt(1);
+
+            $invalid_casts[] = $atomic_type->getId();
+        }
+
+        if ($invalid_casts) {
+            IssueBuffer::maybeAdd(
+                new InvalidCast(
+                    $invalid_casts[0] . ' cannot be cast to int',
+                    new CodeLocation($statements_analyzer->getSource(), $stmt)
+                ),
+                $statements_analyzer->getSuppressedIssues()
+            );
+        } elseif (!empty($risky_cast)) {
+            IssueBuffer::maybeAdd(
+                new RiskyCast(
+                    'Casting ' . $risky_cast[0] . ' to int has possibly unintended value of 0/1',
+                    new CodeLocation($statements_analyzer->getSource(), $stmt)
+                ),
+                $statements_analyzer->getSuppressedIssues()
+            );
+        } elseif ($explicit_cast && !$castable_types) {
+            // todo: emit error here
```

Comment:
Should this be `InvalidCast`?

Reply:
> But this can never be true, therefore I think we can remove this condition?

Probably, but if you like that can wait for #8486. It's probably easier to refactor these separate casts if they're currently as similar as possible.